crates/wasm-mutate/src/mutators/codemotion/ir/mod.rs

```diff
@@ -0,0 +1,501 @@
+
+        let mut frame_stack = vec![(State::Root, None)];
+        let state_stack = RefCell::new(Vec::new());
+        let current_parsing = RefCell::new(Vec::new());
+
+        let push_state = || {
+            state_stack.borrow_mut().push(current_parsing.clone());
+            *current_parsing.borrow_mut() = vec![]
+        };
+        let pop_state = || {
+            state_stack
+                .borrow_mut()
+                .pop()
+                .expect("Missing state")
+                .borrow()
+                .clone()
+        };
+
+        let mut current_code = Range::new(0, 0);
+
+        for (idx, (operator, _)) in operators.iter().enumerate() {
+            match operator {
+                Operator::If { ty } => {
+                    // push current code
+                    if current_code.end != current_code.start {
+                        let id = nodes.borrow().len();
+                        nodes.borrow_mut().push(Node::Code {
+                            range: current_code,
+                        });
+                        current_parsing.borrow_mut().push(id);
+                    }
+                    current_code.start = idx + 1;
+                    current_code.end = idx + 1;
+                    push_state();
+                    frame_stack.push((State::If, Some(ty)));
+                }
+                Operator::Else => {
+                    if current_code.end != current_code.start {
+                        let id = nodes.borrow().len();
+                        nodes.borrow_mut().push(Node::Code {
+                            range: current_code,
+                        });
+                        current_parsing.borrow_mut().push(id);
+                    }
+                    current_code.start = idx + 1;
+                    current_code.end = idx + 1;
+                    push_state();
+                    frame_stack.push((State::Else, None));
+                }
+                Operator::Block { ty } => {
+                    if current_code.end != current_code.start {
+                        let id = nodes.borrow().len();
+                        nodes.borrow_mut().push(Node::Code {
+                            range: current_code,
+                        });
+                        current_parsing.borrow_mut().push(id);
+                    }
+                    current_code.start = idx + 1;
+                    current_code.end = idx + 1;
+                    push_state();
+                    frame_stack.push((State::Block, Some(ty)));
+                }
+                Operator::Loop { ty } => {
+                    if current_code.end != current_code.start {
+                        let id = nodes.borrow().len();
+                        nodes.borrow_mut().push(Node::Code {
+                            range: current_code,
+                        });
+                        current_parsing.borrow_mut().push(id);
+                    }
```

Comment:
This is repeated a bunch. Perhaps it can also be a shared method on the new `ParseContext` struct? It may make sense for the `ParseContext` to contain the `nodes`, `ifs`, `loops`, etc as well and then have a `finish` method for when the parse is done for creating the root node and the final `Ast` and all that.